File: sol/solution/0001-0100/0037/Solution.swift
class Solution {
    func solveSudoku(_ board: inout [[Character]]) {
        _ = solve(&board)
    }
    private func solve(_ b: inout [[Character]]) -> Bool {
        for i in 0..<9 {
            for j in 0..<9 {
                if b[i][j] == "." {
                    for d in 1...9 {
                        let ch = Character(String(d))
                        if isValid(b, i, j, ch) {
                            b[i][j] = ch
                            if solve(&b) { return true }
                            b[i][j] = "."
                        }
                    }
                    return false
                }
            }
        }
        return true
    }
    private func isValid(_ b: [[Character]], _ r: Int, _ c: Int, _ ch: Character) -> Bool {
        for i in 0..<9 {
            if b[r][i] == ch || b[i][c] == ch { return false }
            if b[(r/3)*3 + i/3][(c/3)*3 + i%3] == ch { return false }
        }
        return true
    }
}

File: sol/solution/0001-0100/0037/Solution.cpp
class Solution {
public:
    void solveSudoku(vector<vector<char>>& board) {
        solve(board);
    }
private:
    bool solve(vector<vector<char>>& b) {
        for (int i = 0; i < 9; i++) {
            for (int j = 0; j < 9; j++) {
                if (b[i][j] == '.') {
                    for (char c = '1'; c <= '9'; c++) {
                        if (isValid(b, i, j, c)) {
                            b[i][j] = c;
                            if (solve(b)) return true;
                            b[i][j] = '.';
                        }
                    }
                    return false;
                }
            }
        }
        return true;
    }
    bool isValid(vector<vector<char>>& b, int r, int c, char ch) {
        for (int i = 0; i < 9; i++) {
            if (b[r][i] == ch || b[i][c] == ch) return false;
            if (b[(r/3)*3 + i/3][(c/3)*3 + i%3] == ch) return false;
        }
        return true;
    }
};